File: obs_replaykit/startup.py
"""windows startup integration for launching OBS with ReplayKit."""

from __future__ import annotations

import os
import base64
import json
import subprocess
from pathlib import Path
from typing import Callable, Optional

from .obs import OBS_START_ARGS, find_obs_exe

LogFn = Optional[Callable[[str], None]]

_RUN_KEY = r"Software\Microsoft\Windows\CurrentVersion\Run"
_VALUE_NAME = "OBS ReplayKit"
_SHORTCUT_NAME = "OBS ReplayKit.lnk"


def _startup_folder() -> Path | None:
    appdata = os.environ.get("APPDATA")
    if not appdata:
        return None
    return Path(appdata) / "Microsoft" / "Windows" / "Start Menu" / "Programs" / "Startup"


def _startup_shortcut_path() -> Path | None:
    folder = _startup_folder()
    if folder is None:
        return None
    return folder / _SHORTCUT_NAME


def _remove_legacy_run_value(log: LogFn = None) -> bool:
    import winreg

    try:
        with winreg.CreateKeyEx(winreg.HKEY_CURRENT_USER, _RUN_KEY, 0, winreg.KEY_SET_VALUE) as key:
            try:
                winreg.DeleteValue(key, _VALUE_NAME)
                if log:
                    log("Windows startup: removed legacy registry entry")
            except FileNotFoundError:
                pass
        return True
    except OSError as exc:
        if log:
            log(f"warn: could not remove legacy Windows startup registry entry: {exc}")
        return False


def _create_startup_shortcut(log: LogFn = None) -> bool:
    obs_exe = find_obs_exe()
    shortcut_path = _startup_shortcut_path()
    if obs_exe is None:
        if log:
            log("warn: OBS install not found - Windows startup was not changed")
        return False
    if shortcut_path is None:
        if log:
            log("warn: Windows Startup folder was not found")
        return False

    try:
        shortcut_path.parent.mkdir(parents=True, exist_ok=True)
    except OSError as exc:
        if log:
            log(f"warn: could not create Windows Startup folder: {exc}")
        return False

    script = """
$ErrorActionPreference = 'Stop'
$payload = $env:OBS_REPLAYKIT_STARTUP_SHORTCUT | ConvertFrom-Json
$shortcutPath = [string]$payload.shortcutPath
$targetPath = [string]$payload.targetPath
$arguments = [string]$payload.arguments
$workingDirectory = [string]$payload.workingDirectory
$shell = New-Object -ComObject WScript.Shell
$shortcut = $shell.CreateShortcut($shortcutPath)
$shortcut.TargetPath = $targetPath
$shortcut.Arguments = $arguments
$shortcut.WorkingDirectory = $workingDirectory
$shortcut.IconLocation = "$targetPath,0"
$shortcut.Description = 'Start OBS ReplayKit when Windows signs in.'
$shortcut.Save()
"""
    payload = {
        "shortcutPath": str(shortcut_path),
        "targetPath": str(obs_exe),
        "arguments": OBS_START_ARGS,
        "workingDirectory": str(obs_exe.parent),
    }
    env = os.environ.copy()
    env["OBS_REPLAYKIT_STARTUP_SHORTCUT"] = json.dumps(payload)
    encoded_script = base64.b64encode(script.encode("utf-16le")).decode("ascii")
    try:
        result = subprocess.run(
            [
                "powershell.exe",
                "-NoProfile",
                "-ExecutionPolicy",
                "Bypass",
                "-EncodedCommand",
                encoded_script,
            ],
            capture_output=True,
            text=True,
            timeout=20,
            env=env,
            creationflags=0x08000000,
        )
    except (subprocess.TimeoutExpired, OSError) as exc:
        if log:
            log(f"warn: could not create Windows startup shortcut: {exc}")
        return False

    if result.returncode != 0:
        message = (result.stderr or result.stdout or "").strip().splitlines()
        if log:
            log(f"warn: could not create Windows startup shortcut: {message[0] if message else result.returncode}")
        return False

    if log:
        log(f"Windows startup: created shortcut at {shortcut_path}")
    return True


def _delete_startup_shortcut(log: LogFn = None) -> bool:
    shortcut_path = _startup_shortcut_path()
    if shortcut_path is None:
        return True
    try:
        shortcut_path.unlink(missing_ok=True)
        if log:
            log("Windows startup: removed Startup folder shortcut")
        return True
    except OSError as exc:
        if log:
            log(f"warn: could not remove Windows startup shortcut: {exc}")
        return False


def configure_obs_startup(enabled: bool, log: LogFn = None) -> bool:
    """add or remove OBS ReplayKit from the current users Windows startup apps."""
    if os.name != "nt":
        if log:
            log("warn: Windows startup setting is only available on Windows")
        return False

    if enabled:
        registry_removed = _remove_legacy_run_value(log)
        created = _create_startup_shortcut(log)
        if created:
            registry_removed = _remove_legacy_run_value(log) and registry_removed
            if log:
                log("Windows startup: OBS will start when you sign in")
        return created and registry_removed

    shortcut_removed = _delete_startup_shortcut(log)
    registry_removed = _remove_legacy_run_value(log)
    if log:
        log("Windows startup: off")
    return shortcut_removed and registry_removed

File: ReplayKitSetup/Startup.cs
using System;
using System.Diagnostics;
using System.IO;

namespace ReplayKitSetup
{
    // windows startup integration for launching obs with replaykit. ported from obs_replaykit/startup.py (the already-fixed version -- plain temp .ps1 invoked with -File, not the base64-encoded -EncodedCommand pattern that AV heuristics flag).
    public static class Startup
    {
        private const string ValueName = "OBS ReplayKit";
        private const string ShortcutName = "OBS ReplayKit.lnk";

        private static string StartupFolder()
        {
            string appdata = Config.APPDATA;
            return appdata == null ? null : Path.Combine(appdata, "Microsoft", "Windows", "Start Menu", "Programs", "Startup");
        }

        private static string StartupShortcutPath()
        {
            string folder = StartupFolder();
            return folder == null ? null : Path.Combine(folder, ShortcutName);
        }

        private static bool RemoveLegacyRunValue(Action<string> log)
        {
            try
            {
                using (var key = Microsoft.Win32.Registry.CurrentUser.OpenSubKey(@"Software\Microsoft\Windows\CurrentVersion\Run", writable: true))
                {
                    if (key?.GetValue(ValueName) != null)
                    {
                        key.DeleteValue(ValueName, throwOnMissingValue: false);
                        log?.Invoke("Windows startup: removed legacy registry entry");
                    }
                }
                return true;
            }
            catch (Exception exc) when (exc is System.Security.SecurityException || exc is UnauthorizedAccessException || exc is IOException)
            {
                log?.Invoke("warn: could not remove legacy Windows startup registry entry: " + exc.Message);
                return false;
            }
        }

        private static readonly string ShortcutScript =
            "param([string]$ShortcutPath, [string]$TargetPath, [string]$Arguments, [string]$WorkingDirectory)\r\n" +
            "$ErrorActionPreference = 'Stop'\r\n" +
            "$shell = New-Object -ComObject WScript.Shell\r\n" +
            "$shortcut = $shell.CreateShortcut($ShortcutPath)\r\n" +
            "$shortcut.TargetPath = $TargetPath\r\n" +
            "$shortcut.Arguments = $Arguments\r\n" +
            "$shortcut.WorkingDirectory = $WorkingDirectory\r\n" +
            "$shortcut.IconLocation = \"$TargetPath,0\"\r\n" +
            "$shortcut.Description = 'Start OBS ReplayKit when Windows signs in.'\r\n" +
            "$shortcut.Save()\r\n";

        private static bool CreateStartupShortcut(Action<string> log)
        {
            string obsExe = Obs.FindObsExe();
            string shortcutPath = StartupShortcutPath();
            if (obsExe == null)
            {
                log?.Invoke("warn: OBS install not found - Windows startup was not changed");
                return false;
            }
            if (shortcutPath == null)
            {
                log?.Invoke("warn: Windows Startup folder was not found");
                return false;
            }

            try
            {
                Directory.CreateDirectory(Path.GetDirectoryName(shortcutPath));
            }
            catch (Exception exc) when (exc is IOException || exc is UnauthorizedAccessException)
            {
                log?.Invoke("warn: could not create Windows Startup folder: " + exc.Message);
                return false;
            }

            string scriptPath = Path.Combine(Path.GetTempPath(), "obsreplaykit_startup_" + Guid.NewGuid().ToString("N") + ".ps1");
            try
            {
                File.WriteAllText(scriptPath, ShortcutScript, new System.Text.UTF8Encoding(false));
                Process proc;
                try
                {
                    string args = Win32Args.Build(
                        "-NoProfile", "-ExecutionPolicy", "Bypass", "-File", scriptPath,
                        "-ShortcutPath", shortcutPath, "-TargetPath", obsExe,
                        "-Arguments", Obs.OBS_START_ARGS, "-WorkingDirectory", Path.GetDirectoryName(obsExe));
                    var psi = new ProcessStartInfo("powershell.exe", args)
                    {
                        UseShellExecute = false,
                        CreateNoWindow = true,
                        RedirectStandardOutput = true,
                        RedirectStandardError = true,
                    };
                    proc = Process.Start(psi);
                }
                catch (Exception exc) when (exc is System.ComponentModel.Win32Exception || exc is InvalidOperationException)
                {
                    log?.Invoke("warn: could not create Windows startup shortcut: " + exc.Message);
                    return false;
                }

                string stderr = proc.StandardError.ReadToEnd();
                string stdout = proc.StandardOutput.ReadToEnd();
                proc.WaitForExit(20000);

                if (proc.ExitCode != 0)
                {
                    string message = !string.IsNullOrEmpty(stderr) ? stderr : stdout;
                    string firstLine = string.IsNullOrWhiteSpace(message) ? proc.ExitCode.ToString() : message.Trim().Split('\n')[0];
                    log?.Invoke("warn: could not create Windows startup shortcut: " + firstLine);
                    return false;
                }
            }
            finally
            {
                try { File.Delete(scriptPath); } catch (Exception ex) when (ex is IOException || ex is UnauthorizedAccessException) { }
            }

            log?.Invoke("Windows startup: created shortcut at " + shortcutPath);
            return true;
        }

        private static bool DeleteStartupShortcut(Action<string> log)
        {
            string shortcutPath = StartupShortcutPath();
            if (shortcutPath == null) return true;
            try
            {
                if (File.Exists(shortcutPath)) File.Delete(shortcutPath);
                log?.Invoke("Windows startup: removed Startup folder shortcut");
                return true;
            }
            catch (Exception exc) when (exc is IOException || exc is UnauthorizedAccessException)
            {
                log?.Invoke("warn: could not remove Windows startup shortcut: " + exc.Message);
                return false;
            }
        }

        // add or remove obs replaykit from the current users windows startup apps.
        public static bool ConfigureObsStartup(bool enabled, Action<string> log = null)
        {
            if (enabled)
            {
                bool registryRemoved = RemoveLegacyRunValue(log);
                bool created = CreateStartupShortcut(log);
                if (created)
                {
                    registryRemoved = RemoveLegacyRunValue(log) && registryRemoved;
                    log?.Invoke("Windows startup: OBS will start when you sign in");
                }
                return created && registryRemoved;
            }

            bool shortcutRemoved = DeleteStartupShortcut(log);
            bool regRemoved = RemoveLegacyRunValue(log);
            log?.Invoke("Windows startup: off");
            return shortcutRemoved && regRemoved;
        }
    }
}
